addons/auth_totp_mail_enforce/models/res_users.py

```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import hashlib
+import hmac
+import logging
+import re
+import time
+
+from datetime import datetime, timedelta
+
+from odoo import _, models
+from odoo.exceptions import AccessDenied
+from odoo.http import request
+
+_logger = logging.getLogger(__name__)
+
+TOTP_RATE_LIMITS = {
+    'send_email': (_('authentication mails sent'), 10, 3600),
+    'code_check': (_('code verifications'), 10, 3600),
+}
+
+
+class Users(models.Model):
+    _inherit = 'res.users'
+
+    def _mfa_url(self):
+        r = super()._mfa_url()
+        if r is not None:
+            return r
+        ICP = self.env['ir.config_parameter'].sudo()
+        otp_required = False
+        if ICP.get_param('auth_totp.policy') == 'all_required':
+            otp_required = True
+        elif ICP.get_param('auth_totp.policy') == 'employee_required' and self.env.user.has_groups('base.group_user'):
+            otp_required = True
+        if otp_required:
+            return '/web/login/totp'
+
+    def _totp_check(self, code):
+        self._totp_rate_limit('code_check')
+        user = self.sudo()
+        if user.totp_enabled:
+            return super()._totp_check(code)
+
+        if not user._validate_totp_mail_code(code):
+            _logger.info("2FA check (mail): FAIL for %s %r", self, self.login)
+            raise AccessDenied(_("Verification failed, please double-check the 6-digit code"))
+        _logger.info("2FA check(mail): SUCCESS for %s %r", self, self.login)
+        self._totp_rate_limit_purge('code_check')
+
+    def _get_totp_mail_code(self):
+        self.ensure_one()
+        now = datetime.now()
+        ts = int(datetime.timestamp(now) / 3600)
+        token = f'{self.id}/{self.login}'
+
+        msg = f'{token}{ts}'
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        assert secret, "Mail TOTP protection requires a configured database secret"
+        hm = hmac.new(secret.encode('ascii'), msg.encode('utf-8'), hashlib.sha1).hexdigest()
+        code = str(int(hm, 16))[-6:]
+        expiration = now + timedelta(seconds=3600)
+        return code, expiration
+
+    def _validate_totp_mail_code(self, code):
```

Comment:
can't we reuse the `TOTP` class from `auth_totp`?

Reply:
Will check but that will require some factorization than, I think